model/task/task.go

```diff
@@ -2651,44 +2651,147 @@ func (t *Task) MarkUnscheduled() error {
+}
+
+type taskDependencyUpdate struct {
+	// taskIDs is the list of task IDs whose dependencies should be updated.
+	taskIDs []string
+	// taskDependenciesToUpdate is a mapping of task IDs that need updating to
+	// the dependency IDs that should be updated for that task.
+	taskDependenciesToUpdate map[string][]string
+	// newlyBlockedTaskData contains info on tasks that were not blocked but are
+	// about to be blocked.
+	newlyBlockedTaskData []event.TaskBlockedData
+}
+
+func getTaskDependenciesToUpdate(tasks []Task, dependencyIDSet map[string]struct{}, unattainable bool) taskDependencyUpdate {
 	taskIDs := make([]string, 0, len(tasks))
+	taskDependenciesToUpdate := make(map[string][]string)
 	newlyBlockedTaskData := make([]event.TaskBlockedData, 0, len(tasks))
 	for _, t := range tasks {
-		if !t.Blocked() && unattainable && !t.OverrideDependencies {
+		taskIDs = append(taskIDs, t.Id)
+
+		taskCanBeNewlyBlocked := unattainable && !t.Blocked() && !t.OverrideDependencies
+		var taskBlockedOn string
+		for _, dependsOn := range t.DependsOn {
+			depID := dependsOn.TaskId
+			if _, ok := dependencyIDSet[depID]; !ok {
+				continue
+			}
+
+			taskDependenciesToUpdate[t.Id] = append(taskDependenciesToUpdate[t.Id], depID)
+
+			if taskCanBeNewlyBlocked && taskBlockedOn == "" {
+				// If a task is going to be newly blocked by multiple
+				// dependencies at once, just block it arbitrarily on
+				// the first one.
+				taskBlockedOn = depID
+			}
+		}
+
+		if taskCanBeNewlyBlocked && taskBlockedOn != "" {
 			data := event.TaskBlockedData{
 				ID:        t.Id,
 				Execution: t.Execution,
-				BlockedOn: dependencyID,
+				BlockedOn: taskBlockedOn,
 			}
 			newlyBlockedTaskData = append(newlyBlockedTaskData, data)
 		}
-		taskIDs = append(taskIDs, t.Id)
 	}
 
-	if err := updateAllTasksForMatchingDependencies(ctx, taskIDs, dependencyID, unattainable); err != nil {
-		return nil, err
+	return taskDependencyUpdate{
+		taskIDs:                  taskIDs,
+		taskDependenciesToUpdate: taskDependenciesToUpdate,
+		newlyBlockedTaskData:     newlyBlockedTaskData,
 	}
+}
 
-	event.LogManyTasksBlocked(newlyBlockedTaskData)
+// updateTaskDependenciesChunked updates all task dependencies in toUpdate in
+// chunks of updates. The updates must be updated in smaller chunks to avoid the
+// 16 MB update query size limit - if too many task IDs and dependency IDs are
+// passed into a single query, the query size will be too large and the DB will
+// reject it.
+func updateTaskDependenciesChunked(ctx context.Context, toUpdate taskDependencyUpdate, unattainable bool) error {
+	const taskDependencyUpdateChunkSize = 2000
+	var chunkTaskIDs []string
+	chunkDependencyIDSet := make(map[string]struct{})
 
-	updatedTasks, err := FindAll(db.Query(ByIds(taskIDs)))
-	if err != nil {
-		return nil, errors.Wrap(err, "finding updated tasks")
+	for taskID, dependencyIDs := range toUpdate.taskDependenciesToUpdate {
+		if len(chunkDependencyIDSet) >= taskDependencyUpdateChunkSize {
```

Comment:
Should we do the same thing as the for loop above I suggested?

Reply:
Done.